### PR TITLE
Ensemble + sequence matching components (for review; implemented by @somayeh-h)

### PR DESCRIPTION
**Pull request by Tobias-Fischer**

### CODE REVIEW

File: modular_snn/modular_snn_config_evaluation.py [hunk 1]
```diff
@@ -270,8 +291,10 @@ def ignore_hyperactive_neurons(neuron_spikes, testing_result_monitor, threshold_
                         help='Number of excitatory output neurons. The number of inhibitory neurons are defined the same.')
     parser.add_argument('--threshold_i', type=int, default=0, 
                         help='Threshold value used to ignore the hyperactive neurons.')
+    parser.add_argument('--seed', type=int, default=0,
```
Comment (Tobias-Fischer) [inline modular_snn/modular_snn_config_evaluation.py:299]: Briefly describe what random numbers are being generated (what for)

File: modular_snn/modular_snn_config_evaluation.py [hunk 2]
```diff
@@ -49,17 +49,34 @@
...
+        
+        process(args)
+        
+        
+def process(args):
     
     use_precomputed = False 
     
     merged_path = './outputs/outputs_ne{}_L{}'
+    indices_path = "shuffled_indices_L{}_S{}.npy"
```
Comment (Tobias-Fischer) [inline modular_snn/modular_snn_config_evaluation.py]: Would it be useful if this script still works without shuffling?

File: modular_snn/one_snn_module_processing.py [hunk 3]
```diff
@@ -44,6 +45,10 @@ def main(args):
     
     num_test_labels_base = args.num_test_labels
     offset_after_skip_base = args.offset_after_skip
+    args_multi_path_base = args.multi_path
+    args_epochs_base = args.epochs
+    
+    epoch_step = 10
```
Comment (Tobias-Fischer) [inline modular_snn/one_snn_module_processing.py:52]: Needs some brief comment what this is.

File: non_modular_snn/snn_model.py [hunk 4]
```diff
@@ -697,13 +706,15 @@ def set_mode(mode, modes):
...
+                        help='Set seed for random generator.')
 
     parser.add_argument('--ad_path_test', type=str, default="_test_E{}", 
                         help='Additional string arguments to use for saving test outputs in testing')
-    parser.add_argument('--ad_path', type=str, default="_offset{}")             
+    parser.add_argument('--ad_path', type=str, default="_offset{}_S{}")             
     parser.add_argument('--multi_path', type=str, default="epoch{}_T{}_T{}") 
 
-    parser.add_argument('--process_mode', type=str, choices=["train", "test"], default="test", 
+    parser.add_argument('--process_mode', type=str, choices=["train", "record", "calibrate", "test"], default="test",
```
Comment (Tobias-Fischer) [inline non_modular_snn/snn_model.py:717]: Record and calibrate are non-standard terminology and should be explained, also not contained in the help string currently.

File: process_ensembles.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        summed_rates_all.append(summed_rates_i)
+        
+        gt_labels = np.arange(summed_rates_i.shape[0])
+        rates_matrix = invert_dMat(summed_rates_i)
+        inv_test_results_idx = get_pred_results(rates_matrix)
+        numQ = rates_matrix.shape[1]
+        compute_recall(gt_labels, inv_test_results_idx, numQ, n_values, data_path, name=f"recallAtN_SNN{args.seeds[i]}_SL1")
+        
+    print("Ensembling...")
+    summed_rates_all = np.array(summed_rates_all)
```
Comment (Tobias-Fischer) [inline process_ensembles.py]: Probably would be good to briefly add comments what those variables are, e.g. explaining what the dimensions are.

File: process_seqmatch.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    data_path = os.path.join(args.mainfolder_path, results_path)
+    Path(data_path).mkdir(parents=True, exist_ok=True)
+    
+    sys.stdout = Logger(data_path, logfile_name=f"logfile_seqMatching")
+
+    if args.use_ensemble:
+        summed_rates_i = np.load(os.path.join(args.SNN_data_path, "ensemble_summed_rates.npy"))
+    else:
+        summed_rates_i = np.load(os.path.join(args.SNN_data_path, "summed_rates.npy"))
+        shuffled_indices_path = os.path.join(args.mainfolder_path, f"shuffled_indices_L{args.num_query_imgs}_S{args.seeds}.npy")
```
Comment (Tobias-Fischer) [inline process_seqmatch.py]: What if not shuffling is used? Does it still work?

File: process_seqmatch.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    rates_matrix = invert_dMat(summed_rates_i)
+    inv_test_results_idx = get_pred_results(rates_matrix)
+    
+    n_values = [1, 5, 10, 15, 20, 25]
+    numQ = rates_matrix.shape[1]
+    print("\nseq len: 1")
+    compute_recall(gt_labels, inv_test_results_idx, numQ, n_values, data_path, name=f"recallAtN_SNN{args.seeds}_SL1")
+    
+    print("\nSequence matching with varying sequence lengths:")        
+    seq_len_list = np.array([2, 4, 6, 10])
```
Comment (Tobias-Fischer) [inline process_seqmatch.py]: This should be in argparse.

File: process_seqmatch.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    print("done")
+    return 
+
+
+def compute_dist_matrix_seqslam(rates_matrix, seq_len=5):
+    
+    dist_matrix_numpy = torch.from_numpy(rates_matrix)
+    
+    dist_matrix = dist_matrix_numpy.to('cuda').type(torch.cuda.FloatTensor).unsqueeze(0).unsqueeze(0)
```
Comment (Tobias-Fischer) [inline process_seqmatch.py]: Would be good if this still worked if cuda is not available.

File: tools/data_utils.py [hunk 9]
```diff
@@ -85,86 +85,109 @@ def processImage(img, imWidth, imHeight, num_patches):
...
 
-    return filtered_index
 
+def filter_frames(frames, labels, paths_used, type, shuffled_indices, imgLists, offset_after_skip, num_labels, num_test_labels):
+    
+    # shuffle data 
+    paths_used[:len(shuffled_indices)] = paths_used[shuffled_indices]
+    frames[:len(shuffled_indices)] = frames[shuffled_indices]
+    
+    if len(imgLists) == 2:
```
Comment (Tobias-Fischer) [inline tools/data_utils.py]: What is this special case about?

**Review by Tobias-Fischer — COMMENTED** [review]
Some comments inline.

File: modular_snn/modular_snn_config_evaluation.py (hunk 1 above)
Comment (Somayeh-h) [reply modular_snn/modular_snn_config_evaluation.py:299]: Updated the help description

File: modular_snn/modular_snn_config_evaluation.py (hunk 2 above)
Comment (Somayeh-h) [reply modular_snn/modular_snn_config_evaluation.py]: Enabled all scripts to work with/without shuffling

File: modular_snn/one_snn_module_processing.py (hunk 3 above)
Comment (Somayeh-h) [reply modular_snn/one_snn_module_processing.py:52]: Added comment

File: non_modular_snn/snn_model.py (hunk 4 above)
Comment (Somayeh-h) [reply non_modular_snn/snn_model.py:717]: Added in the help string and relevant README files

File: process_ensembles.py (hunk 5 above)
Comment (Somayeh-h) [reply process_ensembles.py]: Added descriptions

File: process_seqmatch.py (hunk 6 above)
Comment (Somayeh-h) [reply process_seqmatch.py]: Yes, enabled

File: process_seqmatch.py (hunk 7 above)
Comment (Somayeh-h) [reply process_seqmatch.py]: Added to argparse

File: process_seqmatch.py (hunk 8 above)
Comment (Somayeh-h) [reply process_seqmatch.py]: Enabled it to work without cuda too if cuda not is not available

File: tools/data_utils.py (hunk 9 above)
Comment (Somayeh-h) [reply tools/data_utils.py]: Updated function to enabled shuffling of N traverses as opposed to just 2 for the reference dataset.

**Review by Somayeh-h — APPROVED** [review]

**Pull request merged**